{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "225199f6-f3fe-418c-8a76-2afca0240199",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# cleanup blastn of pangenome against the 16 strain ORF files, perc_identity=40\n",
    "# NOTE: some ORFs are not in frame, that's why using blastn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "75df3c6d-503b-4823-999d-822c86268016",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import re\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9bced702-e5fd-4191-833c-2892893cdc30",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# map length of MSY strain ORFs\n",
    "file_path = '../data/16_genomes/orf_fastas'\n",
    "files = files = [ \n",
    " 'orfs_MSY24.fasta',\n",
    " 'orfs_MSY25.fasta',\n",
    " 'orfs_MSY26.fasta',\n",
    " 'orfs_MSY27.fasta',\n",
    " 'orfs_MSY28.fasta',\n",
    " 'orfs_MSY29.fasta',\n",
    " 'orfs_MSY30.fasta',\n",
    " 'orfs_MSY31.fasta',\n",
    " 'orfs_MSY32.fasta',\n",
    " 'orfs_MSY33.fasta',\n",
    " 'orfs_MSY34.fasta',\n",
    " 'orfs_MSY35.fasta',\n",
    " 'orfs_MSY36.fasta',\n",
    " 'orfs_MSY37.fasta',\n",
    " 'orfs_MSY38.fasta',\n",
    " 'orfs_MSY39.fasta'\n",
    "]\n",
    "\n",
    "strain = []\n",
    "orf = []\n",
    "seq = []\n",
    "\n",
    "for file in files:\n",
    "    input_file = f\"{file_path}/{file}\"\n",
    "\n",
    "    for record in SeqIO.parse(input_file, \"fasta\"):\n",
    "        strain.append(re.split('\\.|_', file)[1])\n",
    "        orf.append(record.id)\n",
    "        seq.append(str(record.seq))\n",
    "\n",
    "df_16 = pd.DataFrame({'strain':strain, 'orf':orf, 'seq':seq})\n",
    "\n",
    "df_16['orf_len'] = df_16['seq'].str.len()\n",
    "strain_map = dict(zip(df_16['orf'], df_16['orf_len']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d97568ec-05bd-47ca-9730-6969136dcf98",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# map length of pangenome ORFs \n",
    "input_file = '../data/1011_pangenome/allORFs_pangenome.fasta'\n",
    "handle = []\n",
    "seq = []\n",
    "\n",
    "for record in SeqIO.parse(input_file, \"fasta\"):\n",
    "    handle.append(record.id)\n",
    "    seq.append(str(record.seq))\n",
    "    \n",
    "ref_df = pd.DataFrame({'ref_id':handle,'ref_seq':seq})\n",
    "\n",
    "ref_df['ref_len'] = ref_df.ref_seq.str.len()\n",
    "\n",
    "ref_map = dict(zip(ref_df.ref_id, ref_df.ref_len))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e82fa78a-3290-485f-9163-0168605223bc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# combine all the .tsv files into single .tsv\n",
    "files = [ \n",
    " 'MSY24_pangenome.tsv',\n",
    " 'MSY25_pangenome.tsv',\n",
    " 'MSY26_pangenome.tsv',\n",
    " 'MSY27_pangenome.tsv',\n",
    " 'MSY28_pangenome.tsv',\n",
    " 'MSY29_pangenome.tsv',\n",
    " 'MSY30_pangenome.tsv',\n",
    " 'MSY31_pangenome.tsv',\n",
    " 'MSY32_pangenome.tsv',\n",
    " 'MSY33_pangenome.tsv',\n",
    " 'MSY34_pangenome.tsv',\n",
    " 'MSY35_pangenome.tsv',\n",
    " 'MSY36_pangenome.tsv',\n",
    " 'MSY37_pangenome.tsv',\n",
    " 'MSY38_pangenome.tsv',\n",
    " 'MSY39_pangenome.tsv'\n",
    "]\n",
    "dataframes = []\n",
    "blastn_cols = ['query','target','percent_id','alignment_length','mismatch','gapopen','qstart','qend','tstart','tend','evalue','bitscore']\n",
    "\n",
    "for file in files:\n",
    "    temp_df = pd.read_csv(f\"../data/blastn/{file}\", names=blastn_cols, sep='\\t')\n",
    "    dataframes.append(temp_df)\n",
    "    \n",
    "df = pd.concat(dataframes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "91e1d7ef-c2a8-469d-a787-e2beb5fd41bd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# map the length of each strain ORF\n",
    "df['q_seq_len'] = df['query'].map(strain_map)\n",
    "\n",
    "# map the length of each pangenome ORF\n",
    "df['t_seq_len'] = df['target'].map(ref_map)\n",
    "\n",
    "# new column with target strain name\n",
    "df['q_strain'] = df['query'].str.split('_').str[0]\n",
    "\n",
    "# percent id whole\n",
    "df['percent_whole_id_strain-orf-length'] = (((df.alignment_length) / df.q_seq_len) * df.percent_id/100)\n",
    "df['percent_whole_id_pangenome-orf-length'] = (((df.alignment_length) / df.t_seq_len) * df.percent_id/100)\n",
    "\n",
    "# identify pangenome ORFs that are not found in S288C\n",
    "df['t_orf_n'] = df['target'].str.split('-').str[0].astype(int)\n",
    "df['t_orf_not_s288c'] = df['t_orf_n'] <= 1767\n",
    "\n",
    "df.to_csv('output/blastn_16strainORFs-to-pangenome.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a2d956d6-4824-4154-91a4-51b23c6d39f2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>target</th>\n",
       "      <th>percent_id</th>\n",
       "      <th>alignment_length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>q_seq_len</th>\n",
       "      <th>t_seq_len</th>\n",
       "      <th>q_strain</th>\n",
       "      <th>percent_whole_id_strain-orf-length</th>\n",
       "      <th>percent_whole_id_pangenome-orf-length</th>\n",
       "      <th>t_orf_n</th>\n",
       "      <th>t_orf_not_s288c</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MSY24_gene1_additional.copy</td>\n",
       "      <td>1567-maker-CGV_3-7506</td>\n",
       "      <td>81.791</td>\n",
       "      <td>335</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>441</td>\n",
       "      <td>775</td>\n",
       "      <td>285</td>\n",
       "      <td>1</td>\n",
       "      <td>1.710000e-100</td>\n",
       "      <td>366.0</td>\n",
       "      <td>1910</td>\n",
       "      <td>285</td>\n",
       "      <td>MSY24</td>\n",
       "      <td>0.143455</td>\n",
       "      <td>0.961403</td>\n",
       "      <td>1567</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>MSY24_gene2_dubious</td>\n",
       "      <td>1542-maker-BDM_5-23139</td>\n",
       "      <td>92.178</td>\n",
       "      <td>473</td>\n",
       "      <td>26</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>470</td>\n",
       "      <td>772</td>\n",
       "      <td>308</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>683.0</td>\n",
       "      <td>483</td>\n",
       "      <td>792</td>\n",
       "      <td>MSY24</td>\n",
       "      <td>0.902696</td>\n",
       "      <td>0.550508</td>\n",
       "      <td>1542</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>MSY24_gene3_dubious</td>\n",
       "      <td>1567-maker-CGV_3-7506</td>\n",
       "      <td>80.997</td>\n",
       "      <td>321</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>321</td>\n",
       "      <td>15</td>\n",
       "      <td>285</td>\n",
       "      <td>2.130000e-93</td>\n",
       "      <td>341.0</td>\n",
       "      <td>612</td>\n",
       "      <td>285</td>\n",
       "      <td>MSY24</td>\n",
       "      <td>0.424837</td>\n",
       "      <td>0.912282</td>\n",
       "      <td>1567</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>MSY24_gene4_retained</td>\n",
       "      <td>5919-YLR467W_NumOfGenes_63</td>\n",
       "      <td>99.685</td>\n",
       "      <td>317</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>317</td>\n",
       "      <td>3277</td>\n",
       "      <td>3593</td>\n",
       "      <td>4.220000e-162</td>\n",
       "      <td>568.0</td>\n",
       "      <td>317</td>\n",
       "      <td>5391</td>\n",
       "      <td>MSY24</td>\n",
       "      <td>0.996850</td>\n",
       "      <td>0.058616</td>\n",
       "      <td>5919</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>MSY24_gene5_retained</td>\n",
       "      <td>3309-YEL075W-A_NumOfGenes_2</td>\n",
       "      <td>87.629</td>\n",
       "      <td>291</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>291</td>\n",
       "      <td>354</td>\n",
       "      <td>96</td>\n",
       "      <td>9.690000e-107</td>\n",
       "      <td>383.0</td>\n",
       "      <td>309</td>\n",
       "      <td>612</td>\n",
       "      <td>MSY24</td>\n",
       "      <td>0.825244</td>\n",
       "      <td>0.416667</td>\n",
       "      <td>3309</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           query                       target  percent_id  \\\n",
       "12   MSY24_gene1_additional.copy        1567-maker-CGV_3-7506      81.791   \n",
       "40           MSY24_gene2_dubious       1542-maker-BDM_5-23139      92.178   \n",
       "59           MSY24_gene3_dubious        1567-maker-CGV_3-7506      80.997   \n",
       "86          MSY24_gene4_retained   5919-YLR467W_NumOfGenes_63      99.685   \n",
       "104         MSY24_gene5_retained  3309-YEL075W-A_NumOfGenes_2      87.629   \n",
       "\n",
       "     alignment_length  mismatch  gapopen  qstart  qend  tstart  tend  \\\n",
       "12                335        11        2     441   775     285     1   \n",
       "40                473        26        3       1   470     772   308   \n",
       "59                321        11        2       1   321      15   285   \n",
       "86                317         1        0       1   317    3277  3593   \n",
       "104               291         4        2       1   291     354    96   \n",
       "\n",
       "            evalue  bitscore  q_seq_len  t_seq_len q_strain  \\\n",
       "12   1.710000e-100     366.0       1910        285    MSY24   \n",
       "40    0.000000e+00     683.0        483        792    MSY24   \n",
       "59    2.130000e-93     341.0        612        285    MSY24   \n",
       "86   4.220000e-162     568.0        317       5391    MSY24   \n",
       "104  9.690000e-107     383.0        309        612    MSY24   \n",
       "\n",
       "     percent_whole_id_strain-orf-length  \\\n",
       "12                             0.143455   \n",
       "40                             0.902696   \n",
       "59                             0.424837   \n",
       "86                             0.996850   \n",
       "104                            0.825244   \n",
       "\n",
       "     percent_whole_id_pangenome-orf-length  t_orf_n  t_orf_not_s288c  \n",
       "12                                0.961403     1567             True  \n",
       "40                                0.550508     1542             True  \n",
       "59                                0.912282     1567             True  \n",
       "86                                0.058616     5919            False  \n",
       "104                               0.416667     3309            False  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the top percent_id_whole hits\n",
    "# these are the pangenome hits to each of the 16 strain ORFs\n",
    "top_idx_pan = df.groupby(['q_strain','query'])['percent_whole_id_pangenome-orf-length'].transform(max) == df['percent_whole_id_pangenome-orf-length']\n",
    "top_pan = df[top_idx_pan]\n",
    "top_pan.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1291b372-d557-430c-a44c-44f1c6223461",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>target</th>\n",
       "      <th>percent_id</th>\n",
       "      <th>alignment_length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>q_seq_len</th>\n",
       "      <th>t_seq_len</th>\n",
       "      <th>q_strain</th>\n",
       "      <th>percent_whole_id_strain-orf-length</th>\n",
       "      <th>percent_whole_id_pangenome-orf-length</th>\n",
       "      <th>t_orf_n</th>\n",
       "      <th>t_orf_not_s288c</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MSY24_gene1_additional.copy</td>\n",
       "      <td>5919-YLR467W_NumOfGenes_63</td>\n",
       "      <td>96.126</td>\n",
       "      <td>1910</td>\n",
       "      <td>38</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1910</td>\n",
       "      <td>3518</td>\n",
       "      <td>5391</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3140.0</td>\n",
       "      <td>1910</td>\n",
       "      <td>5391</td>\n",
       "      <td>MSY24</td>\n",
       "      <td>0.961260</td>\n",
       "      <td>0.340569</td>\n",
       "      <td>5919</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>MSY24_gene2_dubious</td>\n",
       "      <td>5919-YLR467W_NumOfGenes_63</td>\n",
       "      <td>99.172</td>\n",
       "      <td>483</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>483</td>\n",
       "      <td>5025</td>\n",
       "      <td>4543</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>854.0</td>\n",
       "      <td>483</td>\n",
       "      <td>5391</td>\n",
       "      <td>MSY24</td>\n",
       "      <td>0.991720</td>\n",
       "      <td>0.088852</td>\n",
       "      <td>5919</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>MSY24_gene3_dubious</td>\n",
       "      <td>5919-YLR467W_NumOfGenes_63</td>\n",
       "      <td>93.791</td>\n",
       "      <td>612</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>612</td>\n",
       "      <td>4242</td>\n",
       "      <td>3667</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>961.0</td>\n",
       "      <td>612</td>\n",
       "      <td>5391</td>\n",
       "      <td>MSY24</td>\n",
       "      <td>0.937910</td>\n",
       "      <td>0.106474</td>\n",
       "      <td>5919</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>MSY24_gene4_retained</td>\n",
       "      <td>5919-YLR467W_NumOfGenes_63</td>\n",
       "      <td>99.685</td>\n",
       "      <td>317</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>317</td>\n",
       "      <td>3277</td>\n",
       "      <td>3593</td>\n",
       "      <td>4.220000e-162</td>\n",
       "      <td>568.0</td>\n",
       "      <td>317</td>\n",
       "      <td>5391</td>\n",
       "      <td>MSY24</td>\n",
       "      <td>0.996850</td>\n",
       "      <td>0.058616</td>\n",
       "      <td>5919</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>MSY24_gene5_retained</td>\n",
       "      <td>5919-YLR467W_NumOfGenes_63</td>\n",
       "      <td>98.969</td>\n",
       "      <td>291</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>291</td>\n",
       "      <td>508</td>\n",
       "      <td>798</td>\n",
       "      <td>2.770000e-145</td>\n",
       "      <td>512.0</td>\n",
       "      <td>309</td>\n",
       "      <td>5391</td>\n",
       "      <td>MSY24</td>\n",
       "      <td>0.932038</td>\n",
       "      <td>0.053422</td>\n",
       "      <td>5919</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           query                      target  percent_id  \\\n",
       "0    MSY24_gene1_additional.copy  5919-YLR467W_NumOfGenes_63      96.126   \n",
       "38           MSY24_gene2_dubious  5919-YLR467W_NumOfGenes_63      99.172   \n",
       "52           MSY24_gene3_dubious  5919-YLR467W_NumOfGenes_63      93.791   \n",
       "86          MSY24_gene4_retained  5919-YLR467W_NumOfGenes_63      99.685   \n",
       "103         MSY24_gene5_retained  5919-YLR467W_NumOfGenes_63      98.969   \n",
       "\n",
       "     alignment_length  mismatch  gapopen  qstart  qend  tstart  tend  \\\n",
       "0                1910        38        1       1  1910    3518  5391   \n",
       "38                483         4        0       1   483    5025  4543   \n",
       "52                612         2        1       1   612    4242  3667   \n",
       "86                317         1        0       1   317    3277  3593   \n",
       "103               291         3        0       1   291     508   798   \n",
       "\n",
       "            evalue  bitscore  q_seq_len  t_seq_len q_strain  \\\n",
       "0     0.000000e+00    3140.0       1910       5391    MSY24   \n",
       "38    0.000000e+00     854.0        483       5391    MSY24   \n",
       "52    0.000000e+00     961.0        612       5391    MSY24   \n",
       "86   4.220000e-162     568.0        317       5391    MSY24   \n",
       "103  2.770000e-145     512.0        309       5391    MSY24   \n",
       "\n",
       "     percent_whole_id_strain-orf-length  \\\n",
       "0                              0.961260   \n",
       "38                             0.991720   \n",
       "52                             0.937910   \n",
       "86                             0.996850   \n",
       "103                            0.932038   \n",
       "\n",
       "     percent_whole_id_pangenome-orf-length  t_orf_n  t_orf_not_s288c  \n",
       "0                                 0.340569     5919            False  \n",
       "38                                0.088852     5919            False  \n",
       "52                                0.106474     5919            False  \n",
       "86                                0.058616     5919            False  \n",
       "103                               0.053422     5919            False  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_idx_strain = df.groupby(['q_strain','query'])['percent_whole_id_strain-orf-length'].transform(max) == df['percent_whole_id_strain-orf-length']\n",
    "top_strain = df[top_idx_strain]\n",
    "top_strain.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9441b0ee-1db1-43e1-a39d-3993c316dda2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3349456, 19)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "341dc687-8d7f-43ff-9a86-08a699d3f2f8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "top_df = top_df.groupby(['query'])['percent_whole_id'].max().reset_index()\n",
    "\n",
    "\n",
    "#top_df.to_csv('output/blastn_16strainORFs-to-pangenome_top-hits.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e9da1a66-930a-4f52-a3fe-0607bb789733",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>percent_whole_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MSY24_ORF10_ORF</td>\n",
       "      <td>0.398627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSY24_ORF11_ORF</td>\n",
       "      <td>0.023904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MSY24_ORF12_ORF</td>\n",
       "      <td>0.058104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSY24_ORF13_ORF</td>\n",
       "      <td>0.100295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSY24_ORF14_ORF</td>\n",
       "      <td>0.020563</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             query  percent_whole_id\n",
       "0  MSY24_ORF10_ORF          0.398627\n",
       "1  MSY24_ORF11_ORF          0.023904\n",
       "2  MSY24_ORF12_ORF          0.058104\n",
       "3  MSY24_ORF13_ORF          0.100295\n",
       "4  MSY24_ORF14_ORF          0.020563"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "20079a2e-a301-48e2-bfb5-bffa381d7629",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'target'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m/opt/homebrew/Caskroom/mambaforge/base/envs/data_env/lib/python3.8/site-packages/pandas/core/indexes/base.py:3802\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3801\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3802\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3803\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/mambaforge/base/envs/data_env/lib/python3.8/site-packages/pandas/_libs/index.pyx:138\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/mambaforge/base/envs/data_env/lib/python3.8/site-packages/pandas/_libs/index.pyx:165\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5745\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5753\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'target'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 6\u001b[0m\n\u001b[1;32m      4\u001b[0m pangenome_set \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m(ref_map\u001b[38;5;241m.\u001b[39mkeys())\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# blastn pangenome ORF hits to strains\u001b[39;00m\n\u001b[0;32m----> 6\u001b[0m hit_set \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m(\u001b[43mtop_df\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mtarget\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mtolist())\n\u001b[1;32m      7\u001b[0m \u001b[38;5;66;03m# unique pangenoem ORFs list\u001b[39;00m\n\u001b[1;32m      8\u001b[0m missing_pangenome \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(pangenome_set \u001b[38;5;241m-\u001b[39m hit_set)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/mambaforge/base/envs/data_env/lib/python3.8/site-packages/pandas/core/frame.py:3807\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3805\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   3806\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 3807\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3808\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   3809\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/mambaforge/base/envs/data_env/lib/python3.8/site-packages/pandas/core/indexes/base.py:3804\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3802\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3803\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m-> 3804\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3805\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3806\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3808\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3809\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'target'"
     ]
    }
   ],
   "source": [
    "# identify unique ORFs that did not have a hit, <40% identity threshold\n",
    "\n",
    "# all pangenome ORFs\n",
    "pangenome_set = set(ref_map.keys())\n",
    "# blastn pangenome ORF hits to strains\n",
    "hit_set = set(top_df['target'].tolist())\n",
    "# unique pangenoem ORFs list\n",
    "missing_pangenome = list(pangenome_set - hit_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d637c8e-e155-4d23-a86a-206c472d16a8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "len(missing_pangenome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d1abb69d-2275-4ee4-a64e-7e6b39cbf8ad",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "filename = \"pangenome_unique-ORFs.fasta\"\n",
    "f = open(f\"output/{filename}\", \"w\")\n",
    "\n",
    "for index,row in ref_df.query('ref_id.isin(@missing_pangenome)').iterrows():\n",
    "    f.write(f\">{row.ref_id}\\n\")\n",
    "    f.write(f\"{row.ref_seq}\\n\")\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28e366cf-c1b3-416b-bc4c-815de7f4d60a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75d9f542-38b6-4172-9276-2b3b5f7f7a64",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe038729-14e1-4939-ad76-030e9d44a6ac",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# blastn 16genomes-to-pangenome\n",
    "# column names: https://www.metagenomics.wiki/tools/blast/blastn-output-format-6\n",
    "file = '../data/blastn/16genomes-to-pangenome.tsv'\n",
    "\n",
    "blastn_cols = ['query','target','percent_id','alignment_length','mismatch','gapopen','qstart','qend','tstart','tend','evalue','bitscore']\n",
    "df = pd.read_csv(file, names=blastn_cols, sep='\\t')\n",
    "\n",
    "\n",
    "\n",
    "df['tseq_len'] = df.target.map(ref_map)\n",
    "\n",
    "# percent id whole\n",
    "df['percent_id_whole'] = (((df.alignment_length) / df.tseq_len) * df.percent_id/100)\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c6282807-bded-41fe-a41c-87a21ea87e04",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# blastn 16genomes-to-pangenome\n",
    "# column names: https://www.metagenomics.wiki/tools/blast/blastn-output-format-6\n",
    "file = '../data/blastn/pangenome-to-16genomes.tsv'\n",
    "\n",
    "blastn_cols = ['query','target','percent_id','alignment_length','mismatch','gapopen','qstart','qend','tstart','tend','evalue','bitscore']\n",
    "df = pd.read_csv(file, names=blastn_cols, sep='\\t')\n",
    "\n",
    "df['qseq_len'] = df['query'].map(ref_map)\n",
    "\n",
    "# percent id whole\n",
    "df['percent_whole_id'] = (((df.alignment_length) / df.qseq_len) * df.percent_id/100)\n",
    "\n",
    "df['orf_n'] = df['query'].str.split('-').str[0].astype(int)\n",
    "\n",
    "df['orf_not_s288c'] = df['orf_n'] <= 1767\n",
    "\n",
    "df['t_strain'] = df['target'].str.split('_').str[0]\n",
    "\n",
    "#df.to_csv('output/pangenome_blastn.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fe95f207-933c-4407-9b64-59f0eaa6e54e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# orfs in the pangenome missing from the 16 strains\n",
    "pangenome_set = set(ref_map.keys())\n",
    "\n",
    "hit_set = set(df['query'].tolist())\n",
    "\n",
    "\n",
    "missing_pangenome = list(pangenome_set - hit_set)\n",
    "\n",
    "for index,row in ref_df.query('ref_id.isin(@missing_pangenome)').head().iterrows():\n",
    "    print(row['ref_id'])\n",
    "\n",
    "filename = \"pangenome_unique.fasta\"\n",
    "f = open(f\"output/{filename}\", \"w\")\n",
    "\n",
    "for index,row in ref_df.query('ref_id.isin(@missing_pangenome)').iterrows():\n",
    "    f.write(f\">{row.ref_id}\\n\")\n",
    "    f.write(f\"{row.ref_seq}\\n\")\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "7fef1c4d-1bc7-40bf-bfd7-1c36a4d9886d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_49110/507743706.py:10: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.\n",
      "  top_df['max_percent'] = top_df.max(axis=1)\n"
     ]
    }
   ],
   "source": [
    "# get the top percent_id_whole\n",
    "top_idx = df.groupby(['t_strain','query'])['percent_whole_id'].transform(max) == df['percent_whole_id']\n",
    "top_df = df[top_idx]\n",
    "\n",
    "top_df = top_df.groupby(['query','t_strain']).percent_whole_id.max().reset_index()\n",
    "top_df = top_df.pivot(index='query',columns='t_strain',values='percent_whole_id')\n",
    "\n",
    "top_df.reset_index(inplace=True)\n",
    "\n",
    "top_df['max_percent'] = top_df.max(axis=1)\n",
    "\n",
    "top_df['orf_not_s288c'] = top_df['query'].str.split('-').str[0].astype(int) <=1767\n",
    "\n",
    "top_df.to_csv('output/pangenome_blastn_top-id.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "f0142b31-4e73-43a3-98fd-a5785df8610a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_87177/2384251810.py:4: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.\n",
      "  percent_match.append((top_df.max(axis=1) >= i).sum())\n"
     ]
    }
   ],
   "source": [
    "# would need to create a sliding window of 1-100, then just get the raw number of orfs that fall into that category (loop?)\n",
    "percent_match = []\n",
    "for i in range(100,0,-1):\n",
    "    percent_match.append((top_df.max(axis=1) >= i).sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "8e2caf4c-3b9d-456a-a3c8-2719286058aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>t_strain</th>\n",
       "      <th>query</th>\n",
       "      <th>MSY24</th>\n",
       "      <th>MSY25</th>\n",
       "      <th>MSY26</th>\n",
       "      <th>MSY27</th>\n",
       "      <th>MSY28</th>\n",
       "      <th>MSY29</th>\n",
       "      <th>MSY30</th>\n",
       "      <th>MSY31</th>\n",
       "      <th>MSY32</th>\n",
       "      <th>MSY33</th>\n",
       "      <th>MSY34</th>\n",
       "      <th>MSY35</th>\n",
       "      <th>MSY36</th>\n",
       "      <th>MSY37</th>\n",
       "      <th>MSY38</th>\n",
       "      <th>MSY39</th>\n",
       "      <th>max_percent</th>\n",
       "      <th>orf_not_s288c</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100-augustus_masked-1375-CFF_4</td>\n",
       "      <td>0.917450</td>\n",
       "      <td>0.916820</td>\n",
       "      <td>0.917450</td>\n",
       "      <td>0.916200</td>\n",
       "      <td>0.916820</td>\n",
       "      <td>0.918070</td>\n",
       "      <td>0.917450</td>\n",
       "      <td>0.916200</td>\n",
       "      <td>0.904940</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.916820</td>\n",
       "      <td>0.916820</td>\n",
       "      <td>0.905570</td>\n",
       "      <td>0.916200</td>\n",
       "      <td>0.917450</td>\n",
       "      <td>0.918070</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000-augustus_masked-CGE_5-5812</td>\n",
       "      <td>0.897838</td>\n",
       "      <td>0.897838</td>\n",
       "      <td>0.896792</td>\n",
       "      <td>0.896264</td>\n",
       "      <td>0.896792</td>\n",
       "      <td>0.897310</td>\n",
       "      <td>0.897310</td>\n",
       "      <td>0.896792</td>\n",
       "      <td>0.894680</td>\n",
       "      <td>0.895208</td>\n",
       "      <td>0.895736</td>\n",
       "      <td>0.895208</td>\n",
       "      <td>0.896792</td>\n",
       "      <td>0.896792</td>\n",
       "      <td>0.896792</td>\n",
       "      <td>0.895736</td>\n",
       "      <td>0.897838</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1001-augustus_masked-CGE_5-5812</td>\n",
       "      <td>0.872035</td>\n",
       "      <td>0.871005</td>\n",
       "      <td>0.872035</td>\n",
       "      <td>0.872035</td>\n",
       "      <td>0.872035</td>\n",
       "      <td>0.867907</td>\n",
       "      <td>0.871005</td>\n",
       "      <td>0.871005</td>\n",
       "      <td>0.868936</td>\n",
       "      <td>0.873064</td>\n",
       "      <td>0.869966</td>\n",
       "      <td>0.872035</td>\n",
       "      <td>0.871005</td>\n",
       "      <td>0.872035</td>\n",
       "      <td>0.873064</td>\n",
       "      <td>0.873064</td>\n",
       "      <td>0.873064</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1002-augustus_masked-CGE_5-5812</td>\n",
       "      <td>0.907029</td>\n",
       "      <td>0.903626</td>\n",
       "      <td>0.907029</td>\n",
       "      <td>0.904763</td>\n",
       "      <td>0.907029</td>\n",
       "      <td>0.897957</td>\n",
       "      <td>0.903626</td>\n",
       "      <td>0.904763</td>\n",
       "      <td>0.907029</td>\n",
       "      <td>0.905891</td>\n",
       "      <td>0.904763</td>\n",
       "      <td>0.900223</td>\n",
       "      <td>0.903626</td>\n",
       "      <td>0.905891</td>\n",
       "      <td>0.903626</td>\n",
       "      <td>0.903633</td>\n",
       "      <td>0.907029</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1008-augustus_masked-AID_2-6438</td>\n",
       "      <td>0.900863</td>\n",
       "      <td>0.897989</td>\n",
       "      <td>0.900863</td>\n",
       "      <td>0.897989</td>\n",
       "      <td>0.900863</td>\n",
       "      <td>0.900863</td>\n",
       "      <td>0.896548</td>\n",
       "      <td>0.897989</td>\n",
       "      <td>0.886492</td>\n",
       "      <td>0.897989</td>\n",
       "      <td>0.897989</td>\n",
       "      <td>0.893674</td>\n",
       "      <td>0.899421</td>\n",
       "      <td>0.899421</td>\n",
       "      <td>0.897990</td>\n",
       "      <td>0.900863</td>\n",
       "      <td>0.900863</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7464</th>\n",
       "      <td>995-augustus_masked-AIE_1-5046</td>\n",
       "      <td>0.760028</td>\n",
       "      <td>0.759214</td>\n",
       "      <td>0.760856</td>\n",
       "      <td>0.761670</td>\n",
       "      <td>0.760856</td>\n",
       "      <td>0.760856</td>\n",
       "      <td>0.761670</td>\n",
       "      <td>0.760852</td>\n",
       "      <td>0.758392</td>\n",
       "      <td>0.760856</td>\n",
       "      <td>0.762489</td>\n",
       "      <td>0.760028</td>\n",
       "      <td>0.758396</td>\n",
       "      <td>0.760856</td>\n",
       "      <td>0.762492</td>\n",
       "      <td>0.760028</td>\n",
       "      <td>0.762492</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7465</th>\n",
       "      <td>996-augustus_masked-AIE_1-5046</td>\n",
       "      <td>0.913180</td>\n",
       "      <td>0.917470</td>\n",
       "      <td>0.913180</td>\n",
       "      <td>0.916400</td>\n",
       "      <td>0.913180</td>\n",
       "      <td>0.913180</td>\n",
       "      <td>0.916400</td>\n",
       "      <td>0.916400</td>\n",
       "      <td>0.916400</td>\n",
       "      <td>0.913180</td>\n",
       "      <td>0.916400</td>\n",
       "      <td>0.911040</td>\n",
       "      <td>0.917470</td>\n",
       "      <td>0.913180</td>\n",
       "      <td>0.912110</td>\n",
       "      <td>0.917470</td>\n",
       "      <td>0.917470</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7466</th>\n",
       "      <td>997-augustus_masked-AIE_1-5046</td>\n",
       "      <td>0.876164</td>\n",
       "      <td>0.873065</td>\n",
       "      <td>0.876164</td>\n",
       "      <td>0.876164</td>\n",
       "      <td>0.876164</td>\n",
       "      <td>0.876164</td>\n",
       "      <td>0.876164</td>\n",
       "      <td>0.876164</td>\n",
       "      <td>0.874098</td>\n",
       "      <td>0.876164</td>\n",
       "      <td>0.873065</td>\n",
       "      <td>0.877197</td>\n",
       "      <td>0.873065</td>\n",
       "      <td>0.876164</td>\n",
       "      <td>0.877197</td>\n",
       "      <td>0.877197</td>\n",
       "      <td>0.877197</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7467</th>\n",
       "      <td>998-augustus_masked-BBM_6-7272</td>\n",
       "      <td>0.941600</td>\n",
       "      <td>0.941600</td>\n",
       "      <td>0.941600</td>\n",
       "      <td>0.944440</td>\n",
       "      <td>0.941600</td>\n",
       "      <td>0.935896</td>\n",
       "      <td>0.941600</td>\n",
       "      <td>0.940170</td>\n",
       "      <td>0.943020</td>\n",
       "      <td>0.941600</td>\n",
       "      <td>0.944440</td>\n",
       "      <td>0.941600</td>\n",
       "      <td>0.941600</td>\n",
       "      <td>0.941600</td>\n",
       "      <td>0.940170</td>\n",
       "      <td>0.941600</td>\n",
       "      <td>0.944440</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7468</th>\n",
       "      <td>999-augustus_masked-CGE_5-5812</td>\n",
       "      <td>0.942715</td>\n",
       "      <td>0.942435</td>\n",
       "      <td>0.945525</td>\n",
       "      <td>0.943835</td>\n",
       "      <td>0.944955</td>\n",
       "      <td>0.945525</td>\n",
       "      <td>0.943835</td>\n",
       "      <td>0.943835</td>\n",
       "      <td>0.944115</td>\n",
       "      <td>0.943835</td>\n",
       "      <td>0.943555</td>\n",
       "      <td>0.943555</td>\n",
       "      <td>0.943275</td>\n",
       "      <td>0.944955</td>\n",
       "      <td>0.942435</td>\n",
       "      <td>0.943275</td>\n",
       "      <td>0.945525</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7469 rows Ã— 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "t_strain                            query     MSY24     MSY25     MSY26  \\\n",
       "0          100-augustus_masked-1375-CFF_4  0.917450  0.916820  0.917450   \n",
       "1         1000-augustus_masked-CGE_5-5812  0.897838  0.897838  0.896792   \n",
       "2         1001-augustus_masked-CGE_5-5812  0.872035  0.871005  0.872035   \n",
       "3         1002-augustus_masked-CGE_5-5812  0.907029  0.903626  0.907029   \n",
       "4         1008-augustus_masked-AID_2-6438  0.900863  0.897989  0.900863   \n",
       "...                                   ...       ...       ...       ...   \n",
       "7464       995-augustus_masked-AIE_1-5046  0.760028  0.759214  0.760856   \n",
       "7465       996-augustus_masked-AIE_1-5046  0.913180  0.917470  0.913180   \n",
       "7466       997-augustus_masked-AIE_1-5046  0.876164  0.873065  0.876164   \n",
       "7467       998-augustus_masked-BBM_6-7272  0.941600  0.941600  0.941600   \n",
       "7468       999-augustus_masked-CGE_5-5812  0.942715  0.942435  0.945525   \n",
       "\n",
       "t_strain     MSY27     MSY28     MSY29     MSY30     MSY31     MSY32  \\\n",
       "0         0.916200  0.916820  0.918070  0.917450  0.916200  0.904940   \n",
       "1         0.896264  0.896792  0.897310  0.897310  0.896792  0.894680   \n",
       "2         0.872035  0.872035  0.867907  0.871005  0.871005  0.868936   \n",
       "3         0.904763  0.907029  0.897957  0.903626  0.904763  0.907029   \n",
       "4         0.897989  0.900863  0.900863  0.896548  0.897989  0.886492   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "7464      0.761670  0.760856  0.760856  0.761670  0.760852  0.758392   \n",
       "7465      0.916400  0.913180  0.913180  0.916400  0.916400  0.916400   \n",
       "7466      0.876164  0.876164  0.876164  0.876164  0.876164  0.874098   \n",
       "7467      0.944440  0.941600  0.935896  0.941600  0.940170  0.943020   \n",
       "7468      0.943835  0.944955  0.945525  0.943835  0.943835  0.944115   \n",
       "\n",
       "t_strain     MSY33     MSY34     MSY35     MSY36     MSY37     MSY38  \\\n",
       "0         1.000000  0.916820  0.916820  0.905570  0.916200  0.917450   \n",
       "1         0.895208  0.895736  0.895208  0.896792  0.896792  0.896792   \n",
       "2         0.873064  0.869966  0.872035  0.871005  0.872035  0.873064   \n",
       "3         0.905891  0.904763  0.900223  0.903626  0.905891  0.903626   \n",
       "4         0.897989  0.897989  0.893674  0.899421  0.899421  0.897990   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "7464      0.760856  0.762489  0.760028  0.758396  0.760856  0.762492   \n",
       "7465      0.913180  0.916400  0.911040  0.917470  0.913180  0.912110   \n",
       "7466      0.876164  0.873065  0.877197  0.873065  0.876164  0.877197   \n",
       "7467      0.941600  0.944440  0.941600  0.941600  0.941600  0.940170   \n",
       "7468      0.943835  0.943555  0.943555  0.943275  0.944955  0.942435   \n",
       "\n",
       "t_strain     MSY39  max_percent  orf_not_s288c  \n",
       "0         0.918070     1.000000           True  \n",
       "1         0.895736     0.897838           True  \n",
       "2         0.873064     0.873064           True  \n",
       "3         0.903633     0.907029           True  \n",
       "4         0.900863     0.900863           True  \n",
       "...            ...          ...            ...  \n",
       "7464      0.760028     0.762492           True  \n",
       "7465      0.917470     0.917470           True  \n",
       "7466      0.877197     0.877197           True  \n",
       "7467      0.941600     0.944440           True  \n",
       "7468      0.943275     0.945525           True  \n",
       "\n",
       "[7469 rows x 19 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a45b4a40-41ba-447d-859c-aa02edf2568e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "data_env",
   "language": "python",
   "name": "data_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
